reporter.go

```diff
@@ -22,39 +22,88 @@ package tally
 
 import "time"
 
-// BaseStatsReporter implements the shared reporter methods
+// BaseStatsReporter implements the shared reporter methods.
 type BaseStatsReporter interface {
 	Capabilities() Capabilities
 	Flush()
 }
 
-// StatsReporter is a backend for Scopes to report metrics to
+// StatsReporter is a backend for Scopes to report metrics to.
 type StatsReporter interface {
 	BaseStatsReporter
 
 	// ReportCounter reports a counter value
-	ReportCounter(name string, tags map[string]string, value int64)
+	ReportCounter(
+		name string,
+		tags map[string]string,
+		value int64,
+	)
 
 	// ReportGauge reports a gauge value
-	ReportGauge(name string, tags map[string]string, value float64)
+	ReportGauge(
+		name string,
+		tags map[string]string,
+		value float64,
+	)
 
 	// ReportTimer reports a timer value
-	ReportTimer(name string, tags map[string]string, interval time.Duration)
+	ReportTimer(
+		name string,
+		tags map[string]string,
+		interval time.Duration,
+	)
+
+	// ReportHistogramValueSamples reports histogram samples for a bucket
+	ReportHistogramValueSamples(
+		name string,
+		tags map[string]string,
+		buckets Buckets,
+		bucketLowerBound,
+		bucketUpperBound float64,
+		samples int64,
+	)
+
+	// ReportHistogramDurationSamples reports histogram samples for a bucket
+	ReportHistogramDurationSamples(
+		name string,
+		tags map[string]string,
+		buckets Buckets,
+		bucketLowerBound,
+		bucketUpperBound time.Duration,
```

Comment:
They can be inferred however during emission time to a non-cached reporter it is not very performant to walk the list of buckets and find the lower bound then format that into a stat name.

Passing these values makes it a relatively fast and straight forward implementation for a reporter like `statsd`:
```go
func (r *cactusStatsReporter) ReportHistogramValueSamples(
	name string,
	tags map[string]string,
	buckets tally.Buckets,
	bucketLowerBound,
	bucketUpperBound float64,
	samples int64,
) {
	r.statter.Inc(
		fmt.Sprintf("%s.%s-%s", name,
			r.valueBucketString(bucketLowerBound),
			r.valueBucketString(bucketUpperBound)),
		samples, r.sampleRate)
}

func (r *cactusStatsReporter) ReportHistogramDurationSamples(
	name string,
	tags map[string]string,
	buckets tally.Buckets,
	bucketLowerBound,
	bucketUpperBound time.Duration,
	samples int64,
) {
	r.statter.Inc(
		fmt.Sprintf("%s.%s-%s", name,
			r.durationBucketString(bucketLowerBound),
			r.durationBucketString(bucketUpperBound)),
		samples, r.sampleRate)
}
```